### PR TITLE
Upload Stories Page

### LINKED ISSUE #207: Story Upload - Frontend
Create the story upload form

Pls note commented changes on figma have not been made yet
Upload form: https://www.figma.com/file/uVqnweL2ApqvPHWzQYmZ1t/Dashboard?node-id=115%3A1347
For context (you can split these up)
Success screen: https://www.figma.com/file/uVqnweL2ApqvPHWzQYmZ1t/Dashboard?node-id=129%3A1016
Confirm exit: https://www.figma.com/file/uVqnweL2ApqvPHWzQYmZ1t/Dashboard?node-id=129%3A569

Upload Images:
  - drag and upload feature
  - investigate how this can be done on the frontend (look into [dropzone](https://www.dropzonejs.com/)) and coordinate with backend to upload image

Required fields and field validation are also required

### PR DESCRIPTION
**Pull request by hanlinc27**
61d13894-ae21-4788-b062-aad5f25b66c4 (1)

For image upload: https://www.npmjs.com/package/material-ui-dropzone

image

### Todo
- test it by hitting the actual upload endpoint with the form data body (shown in console)
- story tag autocomplete 
- confirm exit dialog
- success screen
- styling 

Need to merge #214  first, Closes #207

### CODE REVIEW

**Comment by hanlinc27** [comment]
Sorry, please ignore the large diff in the yarn.lock file 😞 , it should be the most updated w/ dependencies though

File: ui/src/pages/Upload.tsx
Comment (AbhiPrasad) [inline ui/src/pages/Upload.tsx]: What do you think about using `useReducer` instead of `useState` here? https://reactjs.org/docs/hooks-reference.html#usereducer

File: ui/src/App.tsx
Comment (AbhiPrasad) [inline ui/src/App.tsx:35]: Should this be nested under admin? `admin/upload`?

File: ui/src/pages/Upload.tsx
Comment (AbhiPrasad) [reply ui/src/pages/Upload.tsx]: I'm thinking specifically when we have to account for error states (like if a user puts in something into a field thats wrong and we wanna indicate that there is a correct option)

File: ui/src/pages/Upload.tsx
Comment (AbhiPrasad) [inline ui/src/pages/Upload.tsx]: let's open a TODO task to care of error state -> we should render something if we get a request error.

File: ui/src/pages/Upload.tsx
Comment (AbhiPrasad) [inline ui/src/pages/Upload.tsx:117]: Once you're confident with the PR, let's write a quick test for the `onSubmit`.

We can mock out the fetch function with `jest.fn()` and the check to see if the data is POSTed in the correct format.

**Review by AbhiPrasad — COMMENTED** [review]
Some comments I had while skimming the PR, will do a more detailed review after it gets outta WIP

File: ui/package.json
Comment (AbhiPrasad) [inline ui/package.json]: We can :fire? Doesn't seem to be used.

**Comment by AbhiPrasad** [comment]
Looking at this again, I'm of the opinion that we can do the PR in parts. We don't need to get styling + error state + requests perfect this time around, we can iterate on it in the future (easier to review this way as well)

File: ui/src/pages/Upload.tsx
Comment (nimwijetunga) [inline ui/src/pages/Upload.tsx:1]: One high level comment is that some of these field are required while others are not.

The non required fields are: author bio, tags and youtube link all other fields are required

I saw things like the youtube link being required so make sure those match what I said above

Comment (nimwijetunga) [inline ui/src/pages/Upload.tsx:1]: One high level comment is that some of these field are required while others are not.

The **non required** fields are: author bio, tags and youtube link (all other fields are required)

I saw things like the youtube link being required so make sure those match what I said above

**Review by nimwijetunga — CHANGES_REQUESTED** [review]

File: ui/src/pages/Upload.tsx
Comment (hanlinc27) [reply ui/src/pages/Upload.tsx:1]: Putting this here from Pam for reference.

File: ui/src/pages/Upload.tsx
Comment (hanlinc27) [reply ui/src/pages/Upload.tsx]: Ok yeah looking at it a bit more I think it'll be a bit cleaner to useReducer for error states instead of going one by one with the MaterialUI component, also might be useful if we recycle this page.

File: ui/src/App.tsx
Comment (hanlinc27) [reply ui/src/App.tsx:35]: Resolving this, will nest it after Login gets merged.

File: ui/src/pages/Upload.tsx
Comment (hanlinc27) [reply ui/src/pages/Upload.tsx:117]: Going to move this to a separate ticket for now, manually tested the skeleton and it worked.

File: ui/src/pages/Upload.tsx
Comment (AbhiPrasad) [inline ui/src/pages/Upload.tsx:13]: Why use a library vs https://material-ui.com/components/autocomplete/#multiple-values?

File: ui/src/pages/Upload.tsx
Comment (AbhiPrasad) [inline ui/src/pages/Upload.tsx]: Let's take constants out of the function and declare them at the top.

File: ui/src/pages/Upload.tsx
Comment (AbhiPrasad) [inline ui/src/pages/Upload.tsx]: Why this over `new File()`?

File: ui/src/pages/Upload.tsx
Comment (AbhiPrasad) [inline ui/src/pages/Upload.tsx]: Why camel case here?

File: ui/src/data/countries.json
Comment (AbhiPrasad) [inline ui/src/data/countries.json:2]: This is fine for now. Do you think we'll ever need to put this in the backend?

File: ui/src/pages/Upload.tsx
Comment (AbhiPrasad) [inline ui/src/pages/Upload.tsx]: Let's set a filesLimit to make sure only 1 image is uploaded.
```suggestion
          <DropzoneArea
            acceptedFiles={["image/*"]}
            filesLimit={1}
            dropzoneText={"Drag and drop an image here or click"}
            onChange={(files) => {
              setImage(files);
            }}
```

**Review by AbhiPrasad — COMMENTED** [review]
We good to go after these comments.

File: ui/src/pages/Upload.tsx
Comment (hanlinc27) [reply ui/src/pages/Upload.tsx:13]: Yeah I can use that one, was mostly playing around and I misunderstood how tags were being added (I thought we wanted to POST tags automatically every time someone added) so was looking for something that had some more built-in onAddition functions.

File: ui/src/pages/Upload.tsx
Comment (hanlinc27) [reply ui/src/pages/Upload.tsx]: hmm I think I was getting errors that way but I can look again

File: ui/src/pages/Upload.tsx
Comment (hanlinc27) [reply ui/src/pages/Upload.tsx]: for a bit more chaos.

File: ui/src/data/countries.json
Comment (hanlinc27) [reply ui/src/data/countries.json:2]: I don't think a countries/cities list is ever used in the backend since we use Mapbox to query the coordinates

File: ui/src/pages/Upload.tsx
Comment (hanlinc27) [reply ui/src/pages/Upload.tsx]: new File means having to initialize all the empty params as well, something like `new File([""], "filename.txt", {type: "text/plain", lastModified: date})`

File: ui/src/pages/Upload.tsx
Comment (AbhiPrasad) [inline ui/src/pages/Upload.tsx:40]: https://www.youtube.com/watch?v=Gn8PDA3owr0

**Review by AbhiPrasad — APPROVED** [review]
Lets gooooooo

**Review by nimwijetunga — APPROVED** [review]

**Pull request merged**
